pinot-common/src/main/java/org/apache/pinot/common/utils/helix/IdealStateGroupCommit.java

```diff
@@ -124,34 +124,27 @@ public IdealState commit(HelixManager helixManager, String resourceName,
           Entry first = queue._pending.poll();
           processed.add(first);
           String mergedResourceName = first._resourceName;
-          HelixDataAccessor dataAccessor = helixManager.getHelixDataAccessor();
-          PropertyKey idealStateKey = dataAccessor.keyBuilder().idealStates(resourceName);
-          IdealState idealState = dataAccessor.getProperty(idealStateKey);
-
-          // Make a copy of the idealState above to pass it to the updater
-          // NOTE: new IdealState(idealState.getRecord()) does not work because it's shallow copy for map fields and
-          // list fields
-          IdealState idealStateCopy = HelixHelper.cloneIdealState(idealState);
 
-          /**
-           * If the local cache does not contain a value, need to check if there is a
-           * value in ZK; use it as initial value if exists
-           */
-          IdealState updatedIdealState = first._updater.apply(idealStateCopy);
-          first._updatedIdealState = updatedIdealState;
-          Iterator<Entry> it = queue._pending.iterator();
-          while (it.hasNext()) {
-            Entry ent = it.next();
-            if (!ent._resourceName.equals(mergedResourceName)) {
-              continue;
-            }
-            processed.add(ent);
-            updatedIdealState = ent._updater.apply(idealStateCopy);
-            ent._updatedIdealState = updatedIdealState;
-            it.remove();
-          }
-          IdealState finalUpdatedIdealState = updatedIdealState;
-          updateIdealState(helixManager, resourceName, anyIdealState -> finalUpdatedIdealState,
+          updateIdealState(helixManager, mergedResourceName, idealState -> {
+                // Make a copy of the idealState above to pass it to the updater
+                // NOTE: new IdealState(idealState.getRecord()) does not work because it's shallow copy for map fields
+                // and list fields
+                IdealState idealStateCopy = HelixHelper.cloneIdealState(idealState);
+                IdealState updatedIdealState = first._updater.apply(idealStateCopy);
+                first._updatedIdealState = updatedIdealState;
+                Iterator<Entry> it = queue._pending.iterator();
+                while (it.hasNext()) {
+                  Entry ent = it.next();
+                  if (!ent._resourceName.equals(mergedResourceName)) {
+                    continue;
+                  }
+                  processed.add(ent);
+                  updatedIdealState = ent._updater.apply(updatedIdealState);
+                  ent._updatedIdealState = updatedIdealState;
+                  it.remove();
+                }
+                return updatedIdealState;
+              },
               retryPolicy, noChangeOk);
```

Comment:
Not quite related to the fix, but if some max attempt failure happens, we would lost track of the series of IS updates  removed from the queue?

Reply:
If I understand the code+fix correctly,
- We could have multiple threads adding different segments to the idealstate.
- This case is fixed in line 142 by passing `updatedIdealState` to the updater, so that all updates are applied _cumulatively_.
- The first thread that enters this logic may end updating all of the segments that other threads have added to the queue.
- The other threads are busy-waiting, and will find the queue empty when their turn comes (i.e. they set `queue._running` to themselves and get to examine the queue), and will return SUCCESS to the caller (irrespective of whether the updating thread has succeeded or not).

Is my understanding right? If so, we should add a status of some sort to each entry, and have only the thread that enters the entry remove it, so that all threads can return errors. Otherwise, the inbound call (a segment push) that caused the first thread to happen will think that it could not push the segment, and a retry may succeed pushing that segment. The other segment pushers will think that their push succeeded and may never retry.

@dinoocch or @jasperjiaguo can you validate this?

In fact, things may get more interesting if the retries start to get added as new entries ... (can that happen?)

Reply:
@mcvsubbu do you mean in the case of an exception writing to zk?

I'm not fully sure of the intent, but the current code has the following behavior:

* The executing thread will throw the exception
* Any grouped executions would `return null`

Particularly, the code uses `Entry._sent` to indicate an entry was already processed and this is the condition for the while loop. So I believe the worst case is that we might either (a) run one extra iteration of the processing or (b) wait 10ms.

We _do_ lose the exception though in all other threads (I don't know if that is an issue or not though/if null is enough to imply an error).

Reply:
Same question as @jasperjiaguo. If the retry eventually fails, the algorithm would have removed the changes from the queue and we would lose them. Perhaps we need to only remove from queue once we know that the update (or retry) has been successful.